target-component-accessor-compiler/ksp/src/main/kotlin/me/tatarka/inject/target/component/accessor/compiler/ksp/TargetComponentAccessorProcessor.kt

```diff
@@ -0,0 +1,146 @@
+package me.tatarka.inject.target.component.accessor.compiler.ksp
+
+import com.google.devtools.ksp.getVisibility
+import com.google.devtools.ksp.processing.CodeGenerator
+import com.google.devtools.ksp.processing.KSPLogger
+import com.google.devtools.ksp.processing.Resolver
+import com.google.devtools.ksp.processing.SymbolProcessor
+import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
+import com.google.devtools.ksp.processing.SymbolProcessorProvider
+import com.google.devtools.ksp.symbol.KSAnnotated
+import com.google.devtools.ksp.symbol.KSDeclaration
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.google.devtools.ksp.symbol.KSTypeReference
+import com.google.devtools.ksp.symbol.Modifier
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ksp.toKModifier
+import com.squareup.kotlinpoet.ksp.toTypeName
+import com.squareup.kotlinpoet.ksp.writeTo
+import me.tatarka.inject.annotations.TargetComponentAccessor
+
+class TargetComponentAccessorProcessor(
+    private val codeGenerator: CodeGenerator,
+    private val logger: KSPLogger,
+) : SymbolProcessor {
+    @Suppress("LoopWithTooManyJumpStatements")
+    override fun process(resolver: Resolver): List<KSAnnotated> {
+        val symbols = resolver.getSymbolsWithAnnotation(
+            requireNotNull(TargetComponentAccessor::class.qualifiedName),
+        )
+
+        for (symbol in symbols) {
+            if (symbol !is KSFunctionDeclaration) {
+                logger.error("TargetComponentAccessor should be applied to a function")
+                continue
+            }
+
+            val funcName = symbol.simpleName.asString()
+
+            // the generated actual funs will be annotated with TargetComponentAccessor
+            // KSP will process them as well so we need to ignore them
+            if (Modifier.ACTUAL in symbol.modifiers && symbol.findExpects().firstOrNull() != null) {
```

Comment:
> the generated actual funs will be annotated with TargetComponentAccessor

does it need to be?

Reply:
I think it's currently a warning if it isn't but it will be turning into an error in k2

Reply:
oh if expect & actual don't match annotations? Yeah I guess that makes sense

Reply:
Yup